### ISSUE TITLE
Bug: Intellisense shows misleading error when parameter is expected to be a union of an object and a string

### ISSUE THREAD

**Issue by devanshj** [microsoft/TypeScript#45217]
# Bug Report

### 🔎 Search Terms

Intellisense, incorrect function parameter

### 🕗 Version & Regression Information

Tried with v4.3.5

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mkLgC4AoeeAChADdcN94BveDATwAcRaAiAQQ4Bp4AZkiS1MMAJYoA5vAC+8AD71GrdvA4AhDnMXqeASngBeAHzwKSccEKFQkWAmTosOPAHkUEJm4BGAK3AaYjJKaloGZjZOHn4hEWwMCWkdJQjVTi05Q1NzS2sbV2BSNKi9bVl9QkKPL18AsAxilVLucv0gA)

### 💻 Code

```ts
declare const send:
  (event: { type: "A", foo: string } | { type: "B" } | "A") => void

declare const sendOnlyObject:
  (event: { type: "A", foo: string } | { type: "B" }) => void


send({ type: "A" })
// Type '"A"' is not assignable to type '"B"'.(2322)

sendOnlyObject({ type: "A" })
// Argument of type '{ type: "A"; }' is not assignable to parameter of type '{ type: "A"; foo: string; } | { type: "B"; }'.
//   Property 'foo' is missing in type '{ type: "A"; }' but required in type '{ type: "A"; foo: string; }'.(2345)
```

### 🙁 Actual behavior

The error for `send({ type: "A" })` is `Type '"A"' is not assignable to type '"B"'`

### 🙂 Expected behavior

The error should be something like the one for `sendOnlyObject` or at not misleading that the property `type` has to be `"B"` that's what is getting suggested here.

**Comment by DanielRosenwasser** [comment]
It's probably the case that we bail early in `findMatchingDiscriminantType` because of the primitive. If it's not too hard and it's not too expensive, the fix is likely to create a union of only object types, match on the discriminants of the union, and elaborate on *that* type.

**Comment by MonstraG** [comment]
I'm not sure if it's related, but I've just hit TS2345 on this:

```ts
const func = (foo: boolean | "bar") => {}
func(1 != 2 && "foo")
```
with argument to func failing with 
`TS2345: Argument of type '"" | "foo"' is not assignable to parameter of type 'boolean | "bar"'.   Type '""' is not assignable to type 'boolean | "bar"'.`
(this contrived example also gives TS2367 but we don't care about that)

It looks like for some reason, if `|` with a string, first `false` is resolved as `""` instead

**Comment by ldwonday** [comment]
Is the problem fix in v4.5.2?

**Comment by MonstraG** [comment]
Neither OP nor mine is fixed on 4.5.2

**Comment by devanshj** [comment]
This seems to be fixed now... not sure by which PR or in which version... but can be closed nonetheless.
